### PR TITLE
feat(ci): move to GitHub Actions

### PR DESCRIPTION
**Pull request by lidel**
## TLDR

The main value added here:
- secrets stay on github, easier to reason about security
- easier QA and release confidence: every commit produces production ready binaries for every supported platform that we can grab and run instead of building PR locally 

## This PR

- Fixes the CI error on Windows during signing – was (not) fun  – details in https://github.com/ipfs-shipyard/ipfs-desktop/pull/1657#issuecomment-701720613
- Migrates our CI/CD setup to a single GitHub Action
- Removes Travis and Appveyor configuration
- This is first time I work with Github Actions – feedback is really appreciated :pray: 

## Workflow highlights

- the same pipeline runs on macOS, Linux and Windows
- triggered by `push` and `pull_request` events
  - we test original commit as-is + check how target branch would behave after PR is merged
- secrets
  - signing for windows and mac, and notarization for mac
  - exposed only when PR was created by Collaborator and disabled for forks
  - this means we still run tests and build binaries for PRs from contributors, the only difference is those are not signed/notarized
- workflow has two stages: 
  - **test**
    - runs lint, unit tests and end-to-end tests with spectron (linux uses xvfb)
  - **build**
    - starts only if tests are green
      - this is important speedup, because build for linux takes time (we have many package types), and this way we immediatelly see failing tests on linux (without waiting 20 minutes)
    - runs  electron-build pipeline and produces packages for specific platform
    - produced packages
      - are attached to GithHub Action and can be downloaded for inspection / testing
        - see [example here](https://github.com/ipfs-shipyard/ipfs-desktop/actions/runs/283172385) (links to `dist...zip` archives)
      - if build occurred on a tag (`v0.13`) then artifacts are also uploaded and attached to release draft on GitHub
         - this means there is no change to the way we do releases and checklist remains the same
- caching
	- downloads done by npm, electron and electron-builder are cached under key based on OS and hash(`package.json`, `package-lock.json` and `electron-builder.yml`), which saves around a minute per job

cc https://github.com/ipfs-shipyard/ipfs-desktop/issues/1263

### CODE REVIEW

**Comment by lidel** [comment]
PSA: I am getting flashbacks of the usual MS Windows SNAFU. I am going to test doing windows build on Github Actions to see if its Appveyor-specific issue.

**Comment by lidel** [comment]
Right. Github Actions are not executed until merged, so I'll have to do local tests via [nektos/act](https://github.com/nektos/act) before this is ready.  Will post update later.

:hole: :rabbit2:

**Comment by lidel** [comment]
:green_heart: Good news: Github Action works fine, so we can drop Appveyor
:broken_heart: Bad news: we have the same error as on Appveyor :upside_down_face: 

```
SignTool Error: The file is being used by another process.
SignTool Error: An error occurred while attempting to sign: D:\a\ipfs-desktop\ipfs-desktop\dist\win-unpacked\resources\app.asar.unpacked\node_modules\go-ipfs\bin\ipfs.exe

  stackTrace=
Error: Exit code: 1. Command failed: C:\Users\runneradmin\AppData\Local\electron-builder\Cache\winCodeSign\winCodeSign-2.6.0\windows-10\x64\signtool.exe sign /t http://timestamp.digicert.com /f C:\Users\RUNNER~1\AppData\Local\Temp\t-T5G40O\0.p12 /d IPFS Desktop /du ipfs-shipyard/ipfs-desktop /p fd63d34e656e0ee0396638165d2f35ed4c8aeab428560f4a83cc90b737c6ea71 (sha256 hash) /debug D:\a\ipfs-desktop\ipfs-desktop\dist\win-unpacked\resources\app.asar.unpacked\node_modules\go-ipfs\bin\ipfs.exe

SignTool Error: The file is being used by another process.
```

**Comment by lidel** [comment]
### Fixed Windows signing

Good news: I've found and fixed the problem.

The problem was caused  by clash of:
- the fact that every `.exe` bundled with Electron app is signed 
- our switch from `go-ipfs-dep` to `go-ipfs`, which had symlink in `bin/` directory
- how symlinks are represented when unpacked on MS Windows filesystem


That is why singing `node_modules\go-ipfs\go-ipfs\ipfs.exe ` worked 
but `node_modules\go-ipfs\bin\ipfs.exe` caused  the problem.

```
  •   Signing NSIS uninstaller  file=dist\__uninstaller-nsis-ipfs-desktop.exe certificateFile=C:\Users\RUNNER~1\AppData\Local\Temp\t-j1v0tw\0.p12
  • signing         file=dist\IPFS Desktop Setup 0.12.3.exe certificateFile=C:\Users\RUNNER~1\AppData\Local\Temp\t-j1v0tw\0.p12
  • building block map  blockMapFile=dist\IPFS Desktop Setup 0.12.3.exe.blockmap
  • publishing      publisher=Github (owner: ipfs-shipyard, project: ipfs-desktop, version: 0.12.3)
  • uploading       file=IPFS-Desktop-Setup-0.12.3.exe.blockmap provider=GitHub
  • uploading       file=IPFS-Desktop-Setup-0.12.3.exe provider=GitHub
  • overwrite published file  file=IPFS-Desktop-Setup-0.12.3.exe.blockmap reason=already exists on GitHub
  • overwrite published file  file=IPFS-Desktop-Setup-0.12.3.exe reason=already exists on GitHub
  • overwrite published file  file=latest.yml reason=already exists on GitHub
```


### Next: move everything to  Github Action

While debugging the issue I've created Github Action setup (to confirm its Window issue, not AppVeyor), and it could also replace Travis for macOS and Linux, so I'm removing AppVeyor for building Windows binaries in this PR. 

@rafaelramalho19 @jessicaschilling 
The Action I created is easy to extend for supporting macOS and Linux, so I'll try to spend some time on Thursday to push this over the finish line, migrate secrets to github, and then do test release with new setup without Travis.

I'll timebox it to one day to avoid :hole: :rabbit2:, if this takes too much time, we can keep Travis for macOS and Linux, just to unblock the release.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        run: npm ci --prefer-offline --no-audit --progress=false --cache .cache/npm
+
+      - name: Build
+        run: npm run build
+
+      - name: Test
+        run: npm run test
+
+      - name: Test end-to-end
+        # TODO if: !startsWith(matrix.os, 'ubuntu') # xvfb slow on gh, we run test:e2e on Travis for now
```
Comment (rafaelramalho19) [inline .github/workflows/build.yml:49]: Is this something we can check in the test:e2e? Is it only slow for gh or is it also slow for ubuntu users?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        run: npm run test:e2e
+
+      - name: Lint
+        run: npm run lint
+
+      # TODO: run electron builder only on release tag? ${{ startsWith(github.ref, 'refs/tags/v') }}
```
Comment (rafaelramalho19) [inline .github/workflows/build.yml:55]: can you use:
```
on:
  create:
    tags:
      - v*
```

?

File: .github/workflows/build.yml (hunk 1 above)
Comment (lidel) [reply .github/workflows/build.yml]: Unsure why test:e2e fails on Linux at GH. The same code runs fine on Linux on Travis, so I assume something is wrong with the way we run xvfb (we spawn it separately there)

**Comment by lidel** [comment]
Ok, this is ready for review. 

@rafaelramalho19  I migrated secrets and based on the lack of fireworks I believe CI correctly signs Windows and macOS binaries, but would like second set of eyes to confirm by dowloading `.zip` with package for respective platform and smoke-testing  Windows/macOS artifacts attached to this run: https://github.com/ipfs-shipyard/ipfs-desktop/actions/runs/283172385

- [ ] is Windows signed? [dist-windows-latest.zip](https://github.com/ipfs-shipyard/ipfs-desktop/suites/1275756644/artifacts/19704163)
- [ ] is macOS signed and notarized? [dist-macos-latest.zip](https://github.com/ipfs-shipyard/ipfs-desktop/suites/1275756644/artifacts/19704161)

**Comment by jessicaschilling** [comment]
Downloaded from [here](https://github.com/ipfs-shipyard/ipfs-desktop/suites/1275756644/artifacts/19704161), unzipped, got this
image
... and successfully launched.

**Review by jessicaschilling — APPROVED** [review]

**Comment by aschmahmann** [comment]
@lidel seems to be working fine on Windows and looks signed.

**Review by hacdias — APPROVED** [review]
I must say I'm happy we can ditch Appveyor and Travis and put everything together in GitHub actions. Everything looks fine for me! Thanks @lidel

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      - name: Test
+        run: npm run test
+
+      - name: Test end-to-end
+        uses: GabrielBB/xvfb-action@f040be23a619e5ec34116f24098ad3626ceab681 # v1.4
+        with:
+          working-directory: ${{ github.workspace }}
+          run: npm run test:e2e
+
+      - name: Lint
```
Comment (olizilla) [inline .github/workflows/ci.yml:52]: Why lint so late? is deliberate? could go before build to fail fast on trivial errors?

File: test/e2e/launch.e2e.js [hunk 4]
```diff
@@ -27,7 +27,7 @@ async function getPort () {
 // inside of `daemonReady`. To print them, pass DEBUG=true
 
 describe('Application launch', function () {
-  this.timeout(process.env.CI ? 180000 : 60000)
+  this.timeout(60000)
```
Comment (olizilla) [inline test/e2e/launch.e2e.js:30]: Nice! so fast we dont need th special case!

**Review by olizilla — APPROVED** [review]
it is so clear, i love it!  🏆 ✨

**Review by rafaelramalho19 — APPROVED** [review]
:shipit:

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      - name: Test end-to-end
+        if:  ${{ !startsWith(runner.os, 'Linux') }} # xvfb slow on gh, we run test:e2e on Travis for now
+        run: npm run test:e2e
+
+      - name: Lint
+        run: npm run lint
+
+      # TODO: run electron builder only on release tag? ${{ startsWith(github.ref, 'refs/tags/v') }}
+      - name: Build binaries with electron-builder
+        uses: samuelmeuli/action-electron-builder@92327c67bc45ff7c38bf55d8aa8c4d75b7ea38e7 # v1.6.0 but safer than a tag that can be changed
```
[the remark this answers is not in the document]
Comment (lidel) [reply .github/workflows/build.yml]: This is the version of github action, internally it will use the version of `electron-builder` we have in `package.json`. I don't expect the action to be updated unless there is a breaking-change release of `electron-builder` that we want to switch to. 

ps. I've hardened setup and those third party actions are safelisted in https://github.com/ipfs-shipyard/ipfs-desktop/settings/actions, everything else is blocked. So even if someone sneaks different version in a PR, it won't work.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (lidel) [reply .github/workflows/ci.yml:52]: I think it may be just my personal preference: I find it more friendly to do linting later, 
that way we see that contributed PR passes tests and only need linting fixes (which could be trivial), which makes it easier to evaluate contribution.

**Comment by lidel** [comment]
Thank you all for help with reviews and testing :white_check_mark:  

I'm going to merge this and start release dance, create release draft, but **will wait with publishing Desktop release until Monday**, because I'll be AFK this weekend and in my past life I've learned that shipping to production on Friday is not the best idea 🙈 It will also give us weekend and Monday to smoke-test final binaries.

> Perhaps unrelated to this PR there is a file IPFS Desktop Setup 0.12.3.exe.blockmap inside the zip next to the exe. What's this for?

@aschmahmann iiuc when published next to binaries it will be used by autoupdate mechanism for partial / differential updates, to shrink the update payload size.
